pkg/otlp/config.go

```diff
@@ -29,16 +30,34 @@ func portToUint(v int) (port uint, err error) {
 	return
 }
 
-func fromReceiverSectionConfig(cfg config.Config) *colConfig.Map {
-	return colConfig.NewMapFromStringMap(
-		cfg.GetStringMap(config.OTLPReceiverSection),
-	)
+// readConfigSection from a config.Config object.
+func readConfigSection(cfg config.Config, section string) *colConfig.Map {
+	// Viper doesn't work well when getting subsections, since it
+	// ignores environment variables and nil-but-present sections.
+	// To work around this, we do the following two steps:
+
+	// Step one works around https://github.com/spf13/viper/issues/819
+	// If we only had the stuff below, the nil sections would be ignored.
+	// We want to take into account nil-but-present sections.
+	cfgMap := colConfig.NewMapFromStringMap(cfg.GetStringMap(section))
+
+	// Step two works around https://github.com/spf13/viper/issues/1012
+	// we check every key manually, and if it belongs to the OTLP receiver section,
+	// we set it. We need to do this to account for environment variable values.
+	prefix := section + "."
+	for _, key := range cfg.AllKeys() {
+		if strings.HasPrefix(key, prefix) && cfg.IsSet(key) {
```

Comment:
To make things more fun, `AllKeys` returns all known keys, even those that are not set, so I have to check if they are set before setting them in the config map.